tests/unit/test_tasks.py

```diff
@@ -183,6 +183,195 @@ def test_create_stack(self):
                                        username=self.stack_user_name,
                                        pkey=self.stack_key)
 
+    def test_heat_error_on_first_provider(self):
+        # Setup
+        heat = self.get_heat_client_mock()
+        heat.stacks.get.side_effect = [
+            HTTPException,
+            HTTPNotFound,
+            self.stacks["CREATE_COMPLETE"]
+        ]
+        heat.stacks.create.return_value = {"stack": {"id": self.stack_name}}
+
+        # Run
+        res = LaunchStackTask().run(**self.kwargs)
+
+        # Assertions
+        self.assertEqual(res["status"], "CREATE_COMPLETE")
+        self.assertEqual(res["provider"], self.providers[1]["name"])
+        self.assertEqual(self.get_stack("provider"), self.providers[1]["name"])
+        self.assertEqual(res["error_msg"], "")
+
+    def test_heat_error_on_all_providers(self):
+        # Setup
+        heat = self.get_heat_client_mock()
+        heat.stacks.get.side_effect = [
+            HTTPException,
```

Comment:
Instead of testing with generic `HTTPException` each time, perhaps cycle through some of its various subclasses (such as `HTTPBadRequest`, `HTTPForbidden`, `HTTPServiceUnavailable`) so we make sure we handle all of them properly?

Applies to other tests too.

Reply:
All the other exceptions of this class are subclasses of HTTPException, but I agree that it would make for more realistic tests.  On it.

Reply:
See [this fixup](https://github.com/hastexo/hastexo-xblock/pull/83/commits/0b68b4ff1b3ed5f0eeeeebe74af0f97798bd4f16)